File: DSA/SAMAGRA_590014891/Day_24_Question_3.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

struct Node* createNode(int data) {
    struct Node* newNode = (struct Node*) malloc(sizeof(struct Node));
    newNode->data = data;
    newNode->next = NULL;
    return newNode;
}

struct Node* appendNode(struct Node* head, int data) {
    struct Node* newNode = createNode(data);
    if (head == NULL)
        return newNode;

    struct Node* temp = head;
    while (temp->next != NULL)
        temp = temp->next;
    temp->next = newNode;
    return head;
}

struct Node* deleteAtPosition(struct Node* head, int x) {
    if (head == NULL || x <= 0)
        return head;

    if (x == 1) {
        struct Node* temp = head;
        head = head->next;
        free(temp);
        return head;
    }

    struct Node* current = head;
    for (int i = 1; i < x - 1 && current->next != NULL; i++) {
        current = current->next;
    }

    if (current->next == NULL)
        return head;

    struct Node* nodeToDelete = current->next;
    current->next = nodeToDelete->next;
    free(nodeToDelete);

    return head;
}

void printList(struct Node* head) {
    struct Node* temp = head;
    while (temp != NULL) {
        printf("%d", temp->data);
        if (temp->next != NULL)
            printf(" -> ");
        temp = temp->next;
    }
    printf("\n");
}

int main() {
    struct Node* head = NULL;
    int n, x, value;

    printf("Enter number of nodes: ");
    scanf("%d", &n);

    printf("Enter %d node values:\n", n);
    for (int i = 0; i < n; i++) {
        scanf("%d", &value);
        head = appendNode(head, value);
    }

    printf("Enter the position to delete (1-based): ");
    scanf("%d", &x);

    printf("Original List:\n");
    printList(head);

    head = deleteAtPosition(head, x);

    printf("List after deleting position %d:\n", x);
    printList(head);

    return 0;
}

File: DSA/YashKrishanGupta_590014798/Day_24_Question_3.java
import java.util.Scanner;

public class Day_24_Question_1_2_3 {
    public static class Node{
        int data;
        Node next;

        Node(int data){
            this.data=data;
        }
        Node(){

        }
    }

    //Solution-3
    public static Node deleteXNode(Node head, int x){
        if (head == null || x <= 0) return head;
        Node temp = head;
        if(x==1) return head.next;
        int i = 1;
        while(i<x-1 && temp!=null){
            temp = temp.next;
            i++;
        }
        if(temp==null || temp.next == null) return head;
        temp.next = temp.next.next;
        return head;
    }

    public static Node createList(){
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter the number of nodes : ");
        int n = sc.nextInt();
        if(n==0) return null;
        System.out.print("Enter " + n + " values: ");
        int data = sc.nextInt();
        Node head = new Node(data);
        Node temp = head;
        for(int i=1;i<n;i++){
            data = sc.nextInt();
            temp.next = new Node(data);
            temp = temp.next;
        }
        return head;
    }

    public static void printList(Node head){
        Node temp = head;
        while(temp != null){
            System.out.print(temp.data+"->");
            temp = temp.next;
        }
        System.out.println("null");
    }

    public static void main(String[] args) {
        
        Scanner sc = new Scanner(System.in);
        System.out.println("For deleting Xth node");
        Node head = createList();
        System.out.print("Enter X : ");
        int x = sc.nextInt();
        head = deleteXNode(head,x);  // Question3
        printList(head);

    }
}
